index.js

```diff
@@ -15,21 +15,40 @@ const AbortController = require('abort-controller');
 const fetch = require('node-fetch');
 const {FetchError} = fetch;
 
+function getTimeRemaining(retryOptions) {
+    if (retryOptions && retryOptions.startTime && retryOptions.retryMaxDuration) {
+        const millisEllapsed = Date.now() - retryOptions.startTime;        
+        const remaining = retryOptions.retryMaxDuration - millisEllapsed;
+        return Math.max(0, remaining);
+    } else {
+        return Infinity;
+    }
+}
+
+/**
+ * Have we exceeded the max duration for this fetch operation?
+ * @param {*} retryOptions Options including retryMaxDuration and startTime
+ * @returns True if we have a max duration set and it is exceeded, otherwise false
+ */
+function isResponseTimedOut(retryOptions) {
+    return getTimeRemaining(retryOptions) <= 0;
+}
+
 /**
- * Retry
+ * shouldRetry
  * @param {RetryOptions} retryOptions whether or not to retry on all http error codes or just >500
  * @param {Object} error error object if the fetch request returned an error
  * @param {Object} response fetch call response
  * @returns {Boolean} whether or not to retry the request
  */
-function retry(retryOptions, error, response) {
-    if (retryOptions) {
-        const totalMillisToWait = (retryOptions.retryInitialDelay) + (Date.now() - retryOptions.startTime);
-        return ((totalMillisToWait < retryOptions.retryMaxDuration) &&
-            (error !== null || (retryOptions.retryOnHttpResponse && (retryOptions.retryOnHttpResponse(response))))
-        );
+function shouldRetry(retryOptions, error, response) {
+    if (isResponseTimedOut(retryOptions)) {
+        return false;
+    } else if (retryOptions && retryOptions.retryOnHttpResponse) {
```

Comment:
IMO, we should keep the short circuit to explicitly always retry if fetch throws an error:
```
error != null
```
that way we can avoid refactoring the thrown error into an http response: https://github.com/adobe/node-fetch-retry/blob/c5a2d24e67aef27d94566b2f653fdf52463c9b98/index.js#L194

(see here when node-fetch throws an error: https://github.com/node-fetch/node-fetch/blob/main/docs/ERROR-HANDLING.md)

and to be extra safe, we can make sure the response is defined before calling the httpResponse function
```
    } else if (retryOptions && retryOptions.retryOnHttpResponse) {
          return response && retryOptions.retryOnHttpResponse(response);
```

Reply:
OIC, yeah let me clean this up ... I think I can avoid all of that